### PR TITLE
concurrent cdk: improve resource usage and stop waiting on the main thread

### PR DESCRIPTION
**Pull request by girarda**
## What
* A stripe connection has been [experiencing OOM](https://github.com/airbytehq/oncall/issues/3602).
* This was caused by the concurrent_read_processor keeping all partitions in memory

Two more issues were uncovered along the way:
- While the partition enqueuer and the partition reader should backoff, the main thread shouldn't as it should try to process records as fast as possible
  - Instead of backing off on the main thread, we can backoff on the task threads. PartitionEnqueuer and PartitionReader now wait before enqueing elements if there are too many futures in the threadpool's task list 
- futures.exception is a blocking call. since this was called on the main thread, pruning the futures blocked the threads until the futures completed. this can be resolved by checking if a future raised an exception only if it completed

## How
- Instead of keeping all partitions in memory and checking if they are done, only keep the partitions that are currently running in memory and remove them when they are done
- Before putting new items on the queue, check if we're already at the maximum number of futures. If we are, wait. The queue was wrapped in a ThrottledQueue for ease of use

## Recommended reading order
1. `airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py`
2. `airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/thread_pool_manager.py`
3. `airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/throttler.py`
4. `airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/throttled_queue.py`

### CODE REVIEW

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py [hunk 1]
```diff
@@ -700,19 +700,19 @@ def read_records(self, sync_mode: SyncMode, stream_slice: Optional[Mapping[str,
...
-            sync_mode=SyncMode.full_refresh, cursor_field=cursor_field, stream_state=stream_state
-        )
-        for stream_slice in parent_stream_slices:
-            parent_records = self.parent.read_records(
-                sync_mode=SyncMode.full_refresh, cursor_field=cursor_field, stream_slice=stream_slice, stream_state=stream_state
-            )
-            for record in parent_records:
-                self.logger.info(f"Fetching parent stream slices for stream {self.name}.")
-                yield {"parent": record}
+    # def stream_slices(
```
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:703]: the stripe connector was modified to add prints. This had a significant impact on the conenctor's performance because it prints one line per parent record. this should be reverted before updating the connector again

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partition_reader.py [hunk 2]
```diff
@@ -32,6 +37,8 @@ def process_partition(self, partition: Partition) -> None:
         """
         try:
             for record in partition.read():
+                while self._queue.qsize() > self._max_size:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partition_reader.py:40]: this should have a max attempt

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partition_enqueuer.py [hunk 3]
```diff
@@ -34,6 +39,8 @@ def generate_partitions(self, stream: AbstractStream) -> None:
         """
         try:
             for partition in stream.generate_partitions():
+                while self._queue.qsize() >= self._max_size:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partition_enqueuer.py:42]: this should have a max attempt

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/thread_pool_manager.py [hunk 4]
```diff
@@ -36,17 +36,9 @@ def __init__(
...
         self._futures.append(self._threadpool.submit(function, *args))
 
-    def _wait_while_too_many_pending_futures(self, futures: List[Future[Any]]) -> None:
-        # Wait until the number of pending tasks is < self._max_concurrent_tasks
-        while True:
-            self._prune_futures(futures)
-            if len(futures) < self._max_concurrent_tasks:
-                break
-            self._logger.info("Main thread is sleeping because the task queue is full...")
-            time.sleep(self._sleep_time)
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/thread_pool_manager.py:49]: avoid waiting on the main thread

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/thread_pool_manager.py [hunk 5]
```diff
@@ -60,12 +52,12 @@ def _prune_futures(self, futures: List[Future[Any]]) -> None:
 
         for index in reversed(range(len(futures))):
             future = futures[index]
-            optional_exception = future.exception()
-            if optional_exception:
-                exception = RuntimeError(f"Failed reading with error: {optional_exception}")
-                self._stop_and_raise_exception(exception)
 
             if future.done():
+                optional_exception = future.exception()
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/thread_pool_manager.py:57]: only check if the future is already done

Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/thread_pool_manager.py:58]: Worth adding a comment to explain why. Based on the [doc](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.exception), `If the call hasn’t yet completed then this method will wait up to timeout seconds.`

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/types.py [hunk 6]
```diff
@@ -25,4 +26,23 @@ def __init__(self, partition: Partition):
 """
 Typedef representing the items that can be added to the ThreadBasedConcurrentStream
 """
-QueueItem = Union[Record, Partition, PartitionCompleteSentinel, PartitionGenerationCompletedSentinel, Exception]
+QueueItem = Union[Record, Partition, StreamPartition, PartitionCompleteSentinel, PartitionGenerationCompletedSentinel, Exception]
+
+class QueueItemObject:
+    def __init__(self, item: QueueItem):
+        self.value = item
+        self._order = {
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/types.py:34]: nit: Could be a constant

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partition_enqueuer.py (hunk 3 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partition_enqueuer.py:42]: Can we explain why this is needed on top of the priority? It's not so obvious why both would be needed

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partition_enqueuer.py [hunk 7]
```diff
@@ -34,7 +39,9 @@ def generate_partitions(self, stream: AbstractStream) -> None:
         """
         try:
             for partition in stream.generate_partitions():
-                self._queue.put(partition)
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partition_enqueuer.py:37]: It feels like there is some logic associated with how we access the queue. I wonder if it would make sense to regroup this in a class. We have similar logic in PartitionReader for example

**Review by maxi297 — COMMENTED** [review]
In general, the idea of waiting in the child threads instead of the main one makes sense to me. I'll continue thinking about this until our sync

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partition_enqueuer.py (hunk 7 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partition_enqueuer.py:37]: for sure! I actually ran into another bottleneck preventing the sync from succeeding in a reasonable amount of time. I'll clean this up before asking for a formal review

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py [hunk 8]
```diff
@@ -114,11 +118,10 @@ def on_record(self, record: Record) -> Iterable[AirbyteMessage]:
         message = stream_data_to_airbyte_message(record.stream_name, record.data)
         stream = self._stream_name_to_instance[record.stream_name]
 
-        if self._record_counter[stream.name] == 0:
-            self._logger.info(f"Marking stream {stream.name} as RUNNING")
-            yield stream_status_as_airbyte_message(stream.as_airbyte_stream(), AirbyteStreamStatus.RUNNING)
-
         if message.type == MessageType.RECORD:
+            if self._record_counter[stream.name] == 0:
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py:122]: only emit stream status running if the message is a record

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py [hunk 9]
```diff
@@ -164,27 +167,28 @@ def is_done(self) -> bool:
...
+            and all([len(running_partitions) == 0 for running_partitions in self._streams_to_running_partitions.values()])
+            and len(self._streams_done) == len(self._stream_name_to_instance)
         )
 
     def _is_stream_done(self, stream_name: str) -> bool:
-        return (
-            all([p.is_closed() for p in self._streams_to_partitions[stream_name]])
-            and stream_name not in self._streams_currently_generating_partitions
+        return stream_name in self._streams_done or (
+            len(self._streams_to_running_partitions[stream_name]) == 0 and stream_name not in self._streams_currently_generating_partitions
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py:176]: I don't like the condition `len(self._streams_to_running_partitions[stream_name]) == 0 and stream_name not in self._streams_currently_generating_partitions`

This means that the following assertion is valid even though we haven't processed anything:
```
        handler = ConcurrentReadProcessor(...)
        assert handler._is_stream_done(_STREAM_NAME)
```

For me, it either means that the name is not right or the implementation is not right. It feels like it's the later but this would be harder to fix as we would need to check if the stream has had any partition being processed

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/thread_pool_manager.py [hunk 10]
```diff
@@ -35,19 +34,10 @@ def __init__(
         self._futures: List[Future[Any]] = []
 
     def submit(self, function: Callable[..., Any], *args: Any) -> None:
-        # Submit a task to the threadpool, waiting if there are too many pending tasks
-        self._wait_while_too_many_pending_futures(self._futures)
+        # Submit a task to the threadpool, removing completed tasks if there are too many tasks in self._futures.
+        self._prune_futures(self._futures)
         self._futures.append(self._threadpool.submit(function, *args))
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/thread_pool_manager.py:39]: How do we ensure the limit of futures if we don't wait for threads/futures to be completed? Is `self._threadpool` handling this under the hood?

**Review by maxi297 — COMMENTED** [review]
I have a couple of questions before approving but it seems good in general

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py (hunk 9 above)
Comment (clnoll) [reply airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py:176]: This jumped out at me too. It feels like there should be a more direct way to tell if the stream is done, e.g. by confirming that the sentinels have been seen.

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py [hunk 11]
```diff
@@ -55,6 +55,7 @@ def __init__(
         self._slice_logger = slice_logger
         self._message_repository = message_repository
         self._partition_reader = partition_reader
+        self._streams_done: Set[str] = set()
```
[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py:58]: `set` is the actually python type. it didn't use to be a valid annotation but it is in 3.9. reverted to `Set`, which is the type provided by the typing library for consistency

**Comment by alafanechere** [comment]
@girarda I'm updating your branch as some CI pipeline changes have to be propagated

File: airbyte-cdk/python/unit_tests/sources/streams/concurrent/test_concurrent_partition_generator.py [hunk 12]
```diff
@@ -30,10 +31,4 @@ def test_partition_generator(slices):
...
     partition_generator.generate_partitions(stream)
 
-    actual_partitions = []
-    while partition := queue.get(False):
-        if isinstance(partition, PartitionGenerationCompletedSentinel):
-            break
-        actual_partitions.append(partition)
-
-    assert actual_partitions == partitions
+    assert queue.put.has_calls([call(p) for p in partitions] + [call(PartitionGenerationCompletedSentinel(stream))])
```
Comment (girarda) [inline airbyte-cdk/python/unit_tests/sources/streams/concurrent/test_concurrent_partition_generator.py:34]: verify that the partitions and the sentinel were put on the queue

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/types.py [hunk 13]
```diff
@@ -1,7 +1,7 @@
 #
 # Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 #
-
+from queue import Queue
```
Comment (girarda) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/types.py:4]: unused import

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py [hunk 14]
```diff
@@ -95,8 +96,11 @@ def on_partition_complete_sentinel(self, sentinel: PartitionCompleteSentinel) ->
         """
         partition = sentinel.partition
         partition.close()
-        if self._is_stream_done(partition.stream_name()):
-            yield self._on_stream_is_done(partition.stream_name())
+        partitions_running = self._streams_to_running_partitions[partition.stream_name()]
+        if partition in partitions_running:
+            partitions_running.remove(partition)
+            if len(partitions_running) == 0 and partition.stream_name() not in self._streams_currently_generating_partitions:
```
[the remark this answers is not in the document]
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_read_processor.py]: I don't have a use case for needing the order or reading multiple streams with the same name, but we can change this to an array if we have one

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/throttler.py [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+
+class Throttler:
+    """
+    A throttler that waits until the number of concurrent tasks is below a certain threshold.
+    """
+
+    def __init__(self, futures_list: List[Future[Any]], sleep_time: float, max_concurrent_tasks: int):
+        """
+        :param futures_list: The list of futures to monitor
```
Comment (clnoll) [inline airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/throttler.py:15]: Maybe this is obvious, but why do we need to throttle workers enqueuing items based on the number of pending tasks? I'm wondering if we can solve some of these memory issues by setting [`queue.maxsize`](https://docs.python.org/3/library/queue.html#queue.Queue), which prevents items from being added until something else is dequeued, creating a sort of back pressure.

Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/throttler.py:15]: Good question!

The issue I ran into with blocking on the size of the queue is that the main thread will remove elements from the queue before potentially adding them to the list of futures. Since the main thread doesn't wait, it'll be able to remove items from the queue even if the list of futures is large, so the tasks won't wait.

Screenshot 2024-01-16 at 2 16 22 PM

Comment (clnoll) [reply airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/throttler.py:15]: Cool, thanks for the clarification! I think we should document the reason that we need this. Mind adding a docstring on this file?

**Review by clnoll — APPROVED** [review]
Looks good @girarda! Just one small documentation request.

Would you also mind doing some memory profiling of this branch versus master to show that it fixes the problem?

File: airbyte-integrations/connectors/source-stripe/source_stripe/streams.py [hunk 16]
```diff
@@ -707,19 +707,19 @@ def read_records(self, sync_mode: SyncMode, stream_slice: Optional[Mapping[str,
...
```
Comment (girarda) [inline airbyte-integrations/connectors/source-stripe/source_stripe/streams.py:710]: TODO @girarda: revert this change before merging

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/throttler.py [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+        :param sleep_time: How long to sleep if there are too many pending tasks
+        :param max_concurrent_tasks: The maximum number of tasks that can be pending at the same time
+        """
+        self._futures_list = futures_list
+        self._sleep_time = sleep_time
+        self._max_concurrent_tasks = max_concurrent_tasks
+
+    def wait_and_acquire(self) -> None:
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/throttler.py:23]: Should we have unit tests for that?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/throttled_queue.py [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+        """
+        :param queue: The queue to throttle
+        :param throttler: The throttler to use to throttle the queue
+        :param timeout: The timeout to use when getting items from the queue
+        """
+        self._queue = queue
+        self._throttler = throttler
+        self._timeout = timeout
+
+    def put(self, item: QueueItem) -> None:
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/throttled_queue.py:33]: Should we have unit tests for that?

**Review by maxi297 — APPROVED** [review]

**Comment by girarda** [comment]
[Graph from master showing we rapidly hit almost 5GB of memory usage. This is a little misleading as the sync ended up failing due to a failed heartbeat because the main thread was waiting...
](https://app.datadoghq.com/dashboard/b2x-7y7-izf/concurrent-cdk---stripe-resource-usage?fullscreen_end_ts=1705487535390&fullscreen_paused=true&fullscreen_refresh_mode=paused&fullscreen_section=overview&fullscreen_start_ts=1705465680415&fullscreen_widget=3087295086038654&refresh_mode=paused&tpl_var_ab_connection_id%5B0%5D=c903356a-1ce1-489d-9f84-be27c3749ab1&view=spans&from_ts=1705465680415&to_ts=1705487535390&live=false)

Screenshot 2024-01-17 at 12 02 52 PM

On this branch, resource usage is still ~half the peak, but the sync is able to complete.
Screenshot 2024-01-17 at 12 05 09 PM

File: airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/throttled_queue.py (hunk 18 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/partitions/throttled_queue.py:33]: yes, done!

File: airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/throttler.py (hunk 17 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/throttler.py:23]: yes, done!
